direct_kinematics_ros_plugins/src/DirectKinematicsROSPlugin.cc

```diff
@@ -0,0 +1,1245 @@
+  // --------------------------------------- //
+  double target_pitch = this->prev_pitch;
+  int _pitch_cmd_type = _msg->pitch_cmd_type;
+  // 1 : battery position, 2: target once , 3: target servo
+  if (_pitch_cmd_type == 1)  // battery position
+  {
+    double battpos = _msg->target_pitch_value/39.3701*1000;  // [m to inch]
+    double target_pitch_rad
+              = this->f_pitch_battpos_cal_m * battpos + this->f_pitch_battpos_cal_b;
+    target_pitch = target_pitch_rad;
+  }
+  else if (_pitch_cmd_type == 2)  // target once
+  {
+    target_pitch = _msg->target_pitch_value;
+  }
+  else if (_pitch_cmd_type == 3)  // target servo
+  {
+    // nothing now
+  }
+  else if (_pitch_cmd_type == 0)  // nothing
+  {
+    target_pitch = this->prev_pitch;
+  }
+  else
+  {
+    gzmsg << "WRONG PITCH COMMAND TYPE "
+        << "(1 : battery position, 2: target once, 3: target servo)"
+        << std::endl;
+  }
+  this->prev_pitch = target_pitch;
+
+  // --------------------------------------- //
+  // ----------- RUDDER CONTROL ------------ //
+  // --------------------------------------- //
+  double target_yaw;
+  int _rudder_control_mode = _msg->rudder_control_mode;
+  // 1 : control heading, 2: control angle
+  if (_rudder_control_mode == 1)  // control heading
+  {
+    target_yaw = _msg->target_heading;
+  }
+  else if (_rudder_control_mode == 2)  // control angle
+  {
+    int _rudder_angle = _msg->rudder_angle;
+    // Rudder angle control
+    // 1: center, 2: port, 3: staboard, 4: direct
+    double rudderAngleZero = 0.0;
+    double rudderAnglePort = M_PI/6.0;
+    double rudderAngleStbd = -M_PI/6.0;
+    double rudderAngleTarget = _msg->target_rudder_angle;
+    if (_rudder_angle == 1)  // center
+    {
+      target_yaw = this->prev_yaw + rudderAngleZero;
+    }
+    else if (_rudder_angle == 2)  // port
+    {
+      target_yaw = this->prev_yaw + rudderAnglePort;
+    }
+    else if (_rudder_angle == 3)  // staboard
+    {
+      target_yaw = this->prev_yaw + rudderAngleStbd;
+    }
+    else if (_rudder_angle == 4)  // direct
+    {
+      target_yaw = this->prev_yaw + rudderAngleTarget;
+    }
+    else if (_rudder_angle == 0)  // nothing
+    {
+      target_yaw = this->prev_yaw;
+    }
+    else
+    {
+      gzmsg << "WRONG RUDDER ANGLE COMMAND "
+        << "(1: center, 2: port, 3: staboard, 4: direct)"
+        << std::endl;
+    }
+  }
+  else if (_rudder_control_mode == 0)  // nothing
+  {
+    target_yaw = this->prev_yaw;
+  }
+  else
+  {
+    gzmsg << "WRONG RUDDER COMMAND TYPE "
+        << "(1 : control heading, 2: control angle)"
+        << std::endl;
+  }
+  this->prev_yaw = target_yaw;
+
+  // --------------------------------------- //
+  // ------- MOTOR/THRUSTER COMMAND -------- //
+  // --------------------------------------- //
+  int _motor_cmd_type = _msg->motor_cmd_type;
+  double _motor_cmd_value = _msg->target_motor_cmd;
+  // 1 : voltage control, 2: power control
+  if (_motor_cmd_type == 1 || 2)
```

Comment:
You should really refer to these constants by name instead of value. In Python we could do `UwGliderCommand.MOTOR_CMD_VOLTAGE`. I would assume the C++ bindings have something similar.

Reply:
I don't fully understand this comment. Would it cause problem? You mean **int** _motor_cmd_type should be something like **char**?

Reply:
Ha, my original comment was about a style issue, but looking at this line again I see there's an actual bug.

Style issue: There's nothing technically wrong with something like `_motor_cmd_type == 1`, but it's both not obvious what `1` refers to and it has to be kept in sync with the message definition if the value of voltage control ever changes. Something like `_motor_cmd_type == _msg->MOTOR_CMD_VOLTAGE` (no idea if that's the right syntax) is more self documenting and robust to minor changes in the message file.

Bug: This condition always evaluates to true since 2 is always truthy. You probably meant `(_motor_cmd_type == 1 || _motor_cmd_type == 2)`